### DISCUSSION TITLE
Windows GLPI Agent 1.10 Install by MSI // Delete C: Security

### DISCUSSION THREAD

**Discussion by tristangallet** [glpi-project/glpi-agent#745] [Q&A]
### Bug reporting acknowledgment

Yes, I read it

### Professional support

None

### Describe the bug

We're deploying the Windows 1.10 MSI Agent with a Group Policy Object  as usual :
Computer Configuration -> Policies -> Software Settings -> Software installation -> Original Msi on a share

On several machines, no soucy, but on 5 Lenovo Laptops, the drive C has lost all the security, so the user can not see it (Access denied) and the system is very unstable.

On the Windows Event Viewer, we can see that the MSI is deployed, but ended with an error, maybe because the drive C  is touched.

Event Viewer:
Application Managment policy : event 301 : GPO is attributes
Service Control Manager : ID 7045 : Glpi services are installed : 
Un service a été installé sur le système.

Nom du service :  GLPI Agent
Nom du fichier de service :  "C:\Program Files\GLPI-Agent\perl\bin\glpi-agent.exe" -I"C:\Program Files\GLPI-Agent\perl\agent" -I"C:\Program Files\GLPI-Agent\perl\site\lib" -I"C:\Program Files\GLPI-Agent\perl\vendor\lib" -I"C:\Program Files\GLPI-Agent\perl\lib" "C:\Program Files\GLPI-Agent\perl\bin\glpi-win32-service"
Type de service :  service en mode utilisateur
Type de démarrage du service :  Démarrage automatique
Compte de service :  LocalSystem

Application Management Group policy : Erreur %%1603 : Échec de l’installation de l’application GLPI Agent 1.10 de la stratégie 
Application Management Group policy : Erreur %%1603 : Échec de la suppression de l’attribution de l’application GLPI Agent 1.10
Application Management Group policy : Impossible d’appliquer les modifications aux paramètres d’installation du logiciel. Les modifications du logiciel n’ont pas pu être appliquées. Il devrait exister une entrée de journal précédente avec plus de détails.  L’erreur est : %%1603

### To reproduce

We can not reproduce the bug, but we can not deploy on more 4000 machines with this bug.

### Expected behavior

Installation of the GLPI Agent 1.10 without removing the C: Drive's security.

### Operating system

Windows

### GLPI Agent version

v1.10

### GLPI version

10.0.15

### GLPIInventory plugin or other plugin version

GLPI Inventory v1.3.5

**Comment by tristangallet** [comment]
Adding information : Antivirus is Windows Defender and SCCM Client is installed. OS could be Windows 10 or 11.

**Comment by g-bougard** [comment]
Hi @tristangallet 

> ... but on 5 Lenovo Laptops, the drive C has lost all the security, so the user can not see it (Access denied) and the system is very unstable.

What do you mean by `the drive C has lost all the security` ? Do you mean it occurred during installation or the computer can have lost all the C security for another reason ?
Also, by `the system is very unstable`, do you mean it is only unstable since the installation ?

Anyway, can you also provide your installer commandline ?

In general, a `1603` installation failure is due to an unavailable msi-server service. If you're not using [our official vbs](https://glpi-agent.readthedocs.io/en/latest/installation/index.html#large-installations), you should use it as it checks since few version the msi-server service is available.

**Comment by tristangallet** [comment]
Hello, during installation of the GLPI agent, all security rights have been removed from C:
After rights have been removed, people can not log in and use the computer.
We're using Microsoft Policy to deploy the msi, it has been worked since GLPI agent 1.6 without problem.
Using a vbs is not secure because we can not encrypt it.

**Comment by g-bougard** [comment]
> Using a vbs is not secure because we can not encrypt it.

I'm not a MS Policy deployment expert. I just know 70 or 80% of our customers use the vbs without reporting any problem with glpi-agent 1.10.
Also if I understood well, you put the MSI on a shared folder. I imagine this share is secured. Why not putting the vbs in the same place and start it in place of a msiexec command ?

This is up to you to explain us how the installation is processing in you context. We can't guess it. Until you provide more info (like the installer commandline I requested), we can only suppose you're triggering a windows os issue for which we can't do anything. And for now, you provided too poor info, so I can't help to identify what could be the real problem.

A point, you can add an installer option to generate a full installer log: for example `/L*V "C:\glpi-agent-install.log"`, see [msiexec windows commands doc](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/msiexec) for more details.

**Comment by tristangallet** [comment]
We're not using a command but a policy to deploy it.
To configure the package with the good parameters, we have a policy to add the good registry keys.
It's not a security problem, the share is the same and the security is READ for auhtenticated. We can see the log that the Windows policy has been read. I know that this is poor info, it's not logical and the first time i've seen such a bug. I will test the policy this week on more machine to see if it comes again.

**Comment by g-bougard** [comment]
What was the latest version you installed without any error ?

**Comment by tristangallet** [comment]
GLPI Agent 1.9

**Comment by g-bougard** [comment]
> GLPI Agent 1.9

Nothing essential changed between 1.9 MSI & 1.10 MSI. Then as I don't see any issue, I'll move this issue as a Q&A discussion if you have more to share or need some help.

**Comment by nbriche** [comment]
We're seeing the exact same problem migrating from 1.7.3 to 1.11.

On the 7/11 we started deploying on ~400 workstations and servers using a GPO running WPKG, as we've done since time untold.

On the 8/11 morning we had a first impacted workstation; we first concluded it was a weird Windows issue and re-imaged it.

Then the same morning, a second, third and fourth workstation were impacted: no executable could be run, and when trying to access the c: drive we got an "Acces denied" error.

We went into threat assessment, disabled every deployment of every update for every piece of software, isolated the impacted workstations and tried to diagnose the issues. Which was difficult, since the only way to access the files were through our XDR.

Turned out that the ACLs for the C: were wiped, thoroughly enough that we couldn't get the rights to correct the rights.

Eventually, with the help of our cyberdefence cell and Microsoft support, we eventually put aside the threat aspect and managed to reinject the correct rights via our XDR and some powershell black magic.

All deployments were still offline until yesterday, when I enabled 1.11 again. And two hours later I got a new impacted workstation, which deployed 1.11 right before it happened.

**Comment by RaivoMahnovs** [comment]
We're having the same problem migrating from glpi-agent 1.7.1 to 1.14.

Yesterday we started deploying on approx. 600 workstations with CSSM and after couple hours started receiving messages from workstation users, that they cannot open disk C, files and after computer restart, cannot logon. Administrators cannot open disk C even with Admin rights etc.

We also disabled further deployment of glpi-agent. All rights on disk C were removed so to speak. Today we are still having this problem and struggling with finding the right solution.

**Comment by tristangallet** [comment]
Here is some info to recover the C:
Start your Windows in Safe mode or Safe mode with command prompt 

Logon as administrator or administrative right user account 

Go to Run and type = Regedit or if Regedit is not working then start the computer in Safe mode with command prompt Type Regedt32 Go to the following registry key path: HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Policies\System Locate the following REG_DWORD value in Right side pane: ConsentPromptBehaviorAdmin = 0 ConsentPromptBehaviorUser = 0 EnableLUA = 0 Exit from Registry Editor and restart the computer to turn off the UAC 

Reboot the computer and start in Normal mode if working, otherwise in Safe mode 

Go to My computer - Right click on C: drive - Properties - Security - Advance - And now try to change Ownership of the drive and set the permission for it. 

Done! you have now access to C: drive. 

Enable back your UAC as per previous registry value. Go to Run and type Msconfig Go to tools - Change UAC setting - Launch - Set to default - Reboot and Done 

For French, you have to modify according to your groups : 
icacls 'C:' /grant 'AUTORITE NT\Utilisateurs authentifiés:AppendData' icacls 'C:' /grant 'AUTORITE NT\Utilisateurs authentifiés:-536805376' icacls 'C:' /grant 'AUTORITE NT\Système:FullControl' icacls 'C:' /grant 'BUILTIN\Administrateurs:FullControl' icacls 'C:' /grant 'BUILTIN\Utilisateurs:ReadAndExecute, Synchronize'

**Reply by RaivoMahnovs** [reply]
Thank you for answer.

We tried these actions and, unfortunately, it did not resolve all problems. We managed to get back access with Admin profile, but also through Hiren boot CD, not safe mode and it doesn't grant access to all files. Access for some files are still denied and we still strugle with "untying the mystery". 

This problem is still present for GLPI-agent 1.14 and it is not clear for us what causes this, because there are some workstations, where this agent installed succesfully and is not causing any problems.

**Reply by g-bougard** [reply]
What i understand is something could be wrong during installation but only in the context of installation with SCCM.

I don't know how SCCM works. Can anyone explain how the upgrade is setup and what SCCM is expected to do ?

**Reply by RaivoMahnovs** [reply]
From what i know, SCCM is similar to GLPI - it has server and agents installed on workstations. In this SCCM software on server there are collecions of workspaces (grouped as cities or departments) from which we can choose where to deploy software. It works with Active Directory and group policies somehow, but basically SCCM just allows deploying/installing softwares on workspaces through different groups of workstations. It has many options like adding parameter "quiet" to hide setup installation process for users; selecting destination address (in our situation with glpi server) etc., but the difference is that the same setup parameters for glpi-agent worked on some computers, but didn't work on others and there doesn't seem to be any difference between them as in group policies, so elsewhere.

**Reply by g-bougard** [reply]
"deploying software", concerning glpi-agent on windows means you have defined or configured the installation in some way. Nobody still shared how the installation is setup.

I still suggested in this discussion to add a logging option for msiexec... No body tried it. It could be helpfull to obtain such generated log file on a faire case.

One thing the official vbs does is to check MSI service is available. If not, the installation can fail and msiexec tries to rollback the installation. As far i know, this is not always well done. Maybe this is part of the problem as it works on some computers. Maybe something is wrong in the installer, but I can't figure out what with so poorly shared datas on the problem.

**Reply by nbriche** [reply]
Not just SCCM I'm afraid. In our case we use WPKG (https://wpkg.org/), which basically merely wraps around:

msiexec /qn /i "GLPI-Agent-%VERSION%-%ARCH%.msi" SERVER="%serverlist%" ADDLOCAL=ALL DEBUG=1 EXECMODE=1 TAG=XXX RUNNOW=1 AGENTMONITOR=0

and is typically run in the NT Authority\System context, from the logon script section of a computer GPO (although we sometimes run WPKG manually, in which case it's run in the context of whichever Tier-2 admin runs it).

In our case though, as mentioned earlier we managed to "save" the affected hosts by:

0) Use Powershell to get an .xml export of the ACLs from an healthy host. In our case, "acls.xml". I don't have the syntax we used back then, but should be something like `Get-Acl C:\ | Export-Clixml -Path .\acls.xml`. This gets us a ~9.6KB file.

1) Use our XDR's (Palo-Alto Cortex) admin console to open a privileged Powershell. Maybe could get the same thing with safe mode? In our case, that XDR console was the only way we managed it. Now that I think of it, maybe a PXE boot to something like MDT would work?

2) `sfc /scannow`

3) `Set-Acl -Path "c:\" -AclObject (Import-Clixml -Path "acls.xml")`

4) Success. We got back about half a dozen hosts that way.

**Reply by RaivoMahnovs** [reply]
Thank you for your reply. 

Well, we can't even open admin console and enter these commands with admin privileges not in safe mode, not in other way, so, our only option for today is to reinstall workstations completely one by one. We are getting errors, like "Check the spelling of the name"; "Your Start menu isn't working. We'll try to fix it the next time you sign in."; "An attempt was made to reference a token that does not exist." etc.
We can't get any log files from broken workstations either.

I saw in glpi-agent changelogs, that starting version 1.7.2 there were some changes made with privileges:

![image](https://github.com/user-attachments/assets/27348fd2-7309-44c0-9525-c02a6c3b6762)

So i'm assuming that this is the change, that could affect this issue, but i can't figure out what to do if we need to upgrade to future versions.

**Reply by nbriche** [reply]
1.7.3 installs OK here; and my collegues managing some other domains had success with 1.8. I'm not risking anything here and staying on 1.7.3.

Honestly I'm thinking it's a third-party library update that subtly modified the way .msi variables are read, and now under specific (yet unfortunately undefined as of now) circumstances, rights are being removed on C:\ instead of some subfolder.

Do you have a PXE setup? Can you boot on MDT from PXE, and access the console from there?

**Reply by g-bougard** [reply]
IMHO, did someone have the problem making a clean glpi-agent install ? I mean by first running full uninstall of glpi-agent than installing new version. My guess is still something goes wrong during upgrade and MSI decides to revert the upgrade: in that case, sometime things go well, but sometime they go wrong.

**Reply by nbriche** [reply]
As far as I can tell, all the affected clients were indeed upgrades and not clean installs.

**Reply by RaivoMahnovs** [reply]
We upgraded all clients, without uninstalling and installing from scratch. For now we just reverted (reinstalled) affected clients to last state and glpi-agent 1.7.1. We won't upgrade, until this looks safe. :) 
Thanks all for replies, i don't have a clue even after multiple tries of what could cause this problem, but if i'll manage to uncover this, i will post an answer here. Good luck to all!

**Reply by tristangallet** [reply]
When we had the problem, the GLPI client was also being pushed during an installation with SCCM AND with the GPO. The problem was only on Lenovo laptop and not on all, on Windows 10 Wokstation, Windows Server we never had the problem. We have since deactivated all SCCM actions and client 1.14 installed correctly on all, but many machines had received a newer image.

**Reply by nbriche** [reply]
We were hit on both laptops and workstations, but not on servers. I'm not sure why, because they were also targeted for updates, using the same command but with scheduled tasks. I wonder whether that was because the update wasn't done during the bootup phase, or something else.
